

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Getting started (for datasets with unaligned features) &mdash; CAME  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "document", "processHtmlClass": "math|output_area"}}</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Getting started (for datasets with aligned features)" href="getting_started_pipeline_aligned.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> CAME
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../tutorials.html#the-example-data">The example data</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../tutorials.html#getting-started">Getting started</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Getting started (for datasets with unaligned features)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#0-Load-datasets">0 Load datasets</a></li>
<li class="toctree-l4"><a class="reference internal" href="#1-The-default-pipeline-of-CAME">1 The default pipeline of CAME</a></li>
<li class="toctree-l4"><a class="reference internal" href="#2-Further-analysis">2 Further analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="getting_started_pipeline_aligned.html">Getting started (for datasets with aligned features)</a></li>
<li class="toctree-l3"><a class="reference internal" href="load_results.html">Load pre-computed results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API References</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citation.html">Citation</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CAME</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
        
      <li>Getting started (for datasets with unaligned features)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/tut_notebooks/getting_started_pipeline_un.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Getting-started-(for-datasets-with-unaligned-features)">
<h1>Getting started (for datasets with unaligned features)<a class="headerlink" href="#Getting-started-(for-datasets-with-unaligned-features)" title="Permalink to this headline">¶</a></h1>
<p>Import the required packages:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>  <span class="c1"># optional</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>  <span class="c1"># optional</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span>

<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">torch</span>
</pre></div>
</div>
</div>
<p>If you get trouble with installing CAME, you can download the source code from <a class="reference external" href="https://github.com/XingyanLiu/CAME.git">GitHub</a>, and append the path to <code class="docutils literal notranslate"><span class="pre">sys.path</span></code>.</p>
<p>For example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">CAME_ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;path/to/CAME&#39;</span><span class="p">)</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">CAME_ROOT</span><span class="p">))</span>
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">came</span>
<span class="kn">from</span> <span class="nn">came</span> <span class="kn">import</span> <span class="n">pipeline</span><span class="p">,</span> <span class="n">pp</span><span class="p">,</span> <span class="n">pl</span>

<span class="n">ROOT</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>  <span class="c1"># set root</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Using backend: pytorch
</pre></div></div>
</div>
<div class="section" id="0-Load-datasets">
<h2>0 Load datasets<a class="headerlink" href="#0-Load-datasets" title="Permalink to this headline">¶</a></h2>
<div class="section" id="0.1-Load-the-example-datasets">
<h3>0.1 Load the example datasets<a class="headerlink" href="#0.1-Load-the-example-datasets" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">came</span> <span class="kn">import</span> <span class="n">load_example_data</span>

<span class="n">example_data_dict</span> <span class="o">=</span> <span class="n">load_example_data</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">example_data_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<span class="n">adatas</span> <span class="o">=</span> <span class="n">example_data_dict</span><span class="p">[</span><span class="s1">&#39;adatas&#39;</span><span class="p">]</span>
<span class="n">dsnames</span> <span class="o">=</span> <span class="n">example_data_dict</span><span class="p">[</span><span class="s1">&#39;dataset_names&#39;</span><span class="p">]</span>
<span class="n">df_varmap</span> <span class="o">=</span> <span class="n">example_data_dict</span><span class="p">[</span><span class="s1">&#39;varmap&#39;</span><span class="p">]</span>
<span class="n">df_varmap_1v1</span> <span class="o">=</span> <span class="n">example_data_dict</span><span class="p">[</span><span class="s1">&#39;varmap_1v1&#39;</span><span class="p">]</span>

<span class="n">adata_raw1</span><span class="p">,</span> <span class="n">adata_raw2</span> <span class="o">=</span> <span class="n">adatas</span>
<span class="n">key_class1</span> <span class="o">=</span> <span class="n">key_class2</span> <span class="o">=</span> <span class="n">example_data_dict</span><span class="p">[</span><span class="s1">&#39;key_class&#39;</span><span class="p">]</span>

<span class="c1"># setting directory for results</span>
<span class="n">time_tag</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">make_nowtime_tag</span><span class="p">()</span>
<span class="n">resdir</span> <span class="o">=</span> <span class="n">ROOT</span> <span class="o">/</span><span class="s1">&#39;_temp&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">dsnames</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">time_tag</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="n">figdir</span> <span class="o">=</span> <span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;figs&#39;</span>
<span class="n">came</span><span class="o">.</span><span class="n">check_dirs</span><span class="p">(</span><span class="n">figdir</span><span class="p">)</span>  <span class="c1"># check and make the directory</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dict_keys([&#39;adatas&#39;, &#39;varmap&#39;, &#39;varmap_1v1&#39;, &#39;dataset_names&#39;, &#39;key_class&#39;])
a new directory made:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs
</pre></div></div>
</div>
</div>
<div class="section" id="0.2-Load-your-own-datasets">
<h3>0.2 Load your own datasets<a class="headerlink" href="#0.2-Load-your-own-datasets" title="Permalink to this headline">¶</a></h3>
<p>To load your own datasets, see the code example below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># ========= customized paths ==========</span>

<span class="n">dsnames</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;Baron_human&#39;</span><span class="p">,</span> <span class="s1">&#39;Baron_mouse&#39;</span><span class="p">)</span>  <span class="c1"># the dataset names, set by user</span>
<span class="n">dsn1</span><span class="p">,</span> <span class="n">dsn2</span> <span class="o">=</span> <span class="n">dsnames</span>

<span class="n">path_rawdata1</span> <span class="o">=</span> <span class="n">CAME_ROOT</span> <span class="o">/</span> <span class="s1">&#39;came/sample_data/raw-Baron_human.h5ad&#39;</span>
<span class="n">path_rawdata2</span> <span class="o">=</span> <span class="n">CAME_ROOT</span> <span class="o">/</span> <span class="s1">&#39;came/sample_data/raw-Baron_mouse.h5ad&#39;</span>
</pre></div>
</div>
<p>Path to homologous gene mappings, a dataframe with 2-columns, where the first column stores the reference genes and the second column stores the query ones.</p>
<p>Note that <code class="docutils literal notranslate"><span class="pre">path_varmap_1v1</span></code> is optional (can be <code class="docutils literal notranslate"><span class="pre">None</span></code>). If not provided, the 1-to-1 homologous mappings will be inferred from the whole mappings.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path_varmap</span> <span class="o">=</span> <span class="n">CAME_ROOT</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;came/sample_data/gene_matches_human2mouse.csv&#39;</span>
<span class="n">path_varmap_1v1</span> <span class="o">=</span> <span class="n">CAME_ROOT</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;came/sample_data/gene_matches_1v1_human2mouse.csv&#39;</span>
</pre></div>
</div>
<p>Load scRNA-seq datasets.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># ========= load data =========</span>
<span class="n">df_varmap</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path_varmap</span><span class="p">)</span>
<span class="n">df_varmap_1v1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path_varmap_1v1</span><span class="p">)</span> <span class="k">if</span> <span class="n">path_varmap_1v1</span> <span class="k">else</span> <span class="n">came</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">take_1v1_matches</span><span class="p">(</span><span class="n">df_varmap</span><span class="p">)</span>

<span class="n">adata_raw1</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">path_rawdata1</span><span class="p">)</span>
<span class="n">adata_raw2</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">path_rawdata2</span><span class="p">)</span>
<span class="n">adatas</span> <span class="o">=</span> <span class="p">[</span><span class="n">adata_raw1</span><span class="p">,</span> <span class="n">adata_raw2</span><span class="p">]</span>
</pre></div>
</div>
<p>Sepcifiy the column names of the cell-type labels, where <code class="docutils literal notranslate"><span class="pre">key_class1</span></code> is for reference data, and <code class="docutils literal notranslate"><span class="pre">key_class2</span></code> is for query data. If there aren’t any cell-type or clustering labels for the query cells, you can set <code class="docutils literal notranslate"><span class="pre">key_class=None</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">key_class1</span> <span class="o">=</span> <span class="s1">&#39;cell_ontology_class&#39;</span>  <span class="c1"># set by user</span>
<span class="n">key_class2</span> <span class="o">=</span> <span class="s1">&#39;cell_ontology_class&#39;</span>  <span class="c1"># set by user</span>
</pre></div>
</div>
<p>Setting directory for results</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">time_tag</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">make_nowtime_tag</span><span class="p">()</span>
<span class="n">resdir</span> <span class="o">=</span> <span class="n">ROOT</span> <span class="o">/</span><span class="s1">&#39;_temp&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">dsnames</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">time_tag</span><span class="si">}</span><span class="s1">&#39;</span>  <span class="c1"># set by user</span>
<span class="n">figdir</span> <span class="o">=</span> <span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;figs&#39;</span>
<span class="n">came</span><span class="o">.</span><span class="n">check_dirs</span><span class="p">(</span><span class="n">figdir</span><span class="p">)</span>  <span class="c1"># check and make the directory</span>
</pre></div>
</div>
<p>Filtering genes (a preprocessing step, optional)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_raw1</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata_raw2</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="0.3-Inspect-the-compositions-of-different-classes">
<h3>0.3 Inspect the compositions of different classes<a class="headerlink" href="#0.3-Inspect-the-compositions-of-different-classes" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># Inspect classes</span>
<span class="k">if</span> <span class="n">key_class2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
    <span class="n">group_counts_ori</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">adata_raw1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">key_class1</span><span class="p">]),</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">adata_raw2</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">key_class2</span><span class="p">]),</span>
    <span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dsnames</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">group_counts_ori</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">adata_raw1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">key_class1</span><span class="p">])</span>

<span class="n">group_counts_ori</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Baron_human</th>
      <th>Baron_mouse</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>B cell</th>
      <td>NaN</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>Schwann cell</th>
      <td>13.0</td>
      <td>6.0</td>
    </tr>
    <tr>
      <th>T cell</th>
      <td>7.0</td>
      <td>7.0</td>
    </tr>
    <tr>
      <th>endothelial cell</th>
      <td>252.0</td>
      <td>139.0</td>
    </tr>
    <tr>
      <th>leukocyte</th>
      <td>NaN</td>
      <td>8.0</td>
    </tr>
    <tr>
      <th>macrophage</th>
      <td>55.0</td>
      <td>36.0</td>
    </tr>
    <tr>
      <th>mast cell</th>
      <td>25.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>pancreatic A cell</th>
      <td>2326.0</td>
      <td>191.0</td>
    </tr>
    <tr>
      <th>pancreatic D cell</th>
      <td>601.0</td>
      <td>218.0</td>
    </tr>
    <tr>
      <th>pancreatic PP cell</th>
      <td>255.0</td>
      <td>41.0</td>
    </tr>
    <tr>
      <th>pancreatic acinar cell</th>
      <td>958.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>pancreatic ductal cell</th>
      <td>1077.0</td>
      <td>275.0</td>
    </tr>
    <tr>
      <th>pancreatic epsilon cell</th>
      <td>18.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>pancreatic stellate cell</th>
      <td>457.0</td>
      <td>61.0</td>
    </tr>
    <tr>
      <th>type B pancreatic cell</th>
      <td>2525.0</td>
      <td>894.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
</div>
<div class="section" id="1-The-default-pipeline-of-CAME">
<h2>1 The default pipeline of CAME<a class="headerlink" href="#1-The-default-pipeline-of-CAME" title="Permalink to this headline">¶</a></h2>
<p>Parameter settings.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># The numer of training epochs</span>
<span class="c1"># (a recommended setting is 200-400 for whole-graph training, and 80-200 for sub-graph training)</span>
<span class="n">n_epochs</span> <span class="o">=</span> <span class="mi">300</span>

<span class="c1"># The training batch size</span>
<span class="c1"># When the GPU memory is limited, set 4096 or more if possible.</span>
<span class="n">batch_size</span> <span class="o">=</span> <span class="kc">None</span>

<span class="c1"># The number of epochs to skip for checkpoint backup</span>
<span class="n">n_pass</span> <span class="o">=</span> <span class="mi">100</span>

<span class="c1"># Whether to use the single-cell networks</span>
<span class="n">use_scnets</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># The number of top DEGs to take as the node-features of each cells.</span>
<span class="c1"># You set it 70-100 for distant species pairs.</span>
<span class="n">ntop_deg</span> <span class="o">=</span> <span class="mi">50</span>

<span class="c1"># The number of top DEGs to take as the graph nodes, which can be directly displayed on the UMAP plot.</span>
<span class="n">ntop_deg_nodes</span> <span class="o">=</span> <span class="mi">50</span>
<span class="c1"># The source of the node genes; use both DEGs and HVGs by default</span>
<span class="n">node_source</span> <span class="o">=</span> <span class="s1">&#39;deg,hvg&#39;</span>

<span class="c1"># Whether to take into account the non-1v1 variables as the node features.</span>
<span class="n">keep_non1v1_feats</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
<p>💡 An alternative choice for node genes is using only the DEGs:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ntop_deg_nodes</span> <span class="o">=</span> <span class="mi">200</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">node_source</span> <span class="o">=</span> <span class="s1">&#39;deg&#39;</span>
</pre></div>
</div>
<p>⚠️ Note that if most of the homologies are non-1v1, better set <code class="docutils literal notranslate"><span class="pre">keep_non1v1_feats</span></code> as <code class="docutils literal notranslate"><span class="pre">True</span></code>, and increase the number of DEGs used as features <code class="docutils literal notranslate"><span class="pre">ntop_deg</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">came_inputs</span><span class="p">,</span> <span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">)</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">preprocess_unaligned</span><span class="p">(</span>
    <span class="n">adatas</span><span class="p">,</span>
    <span class="n">key_class</span><span class="o">=</span><span class="n">key_class1</span><span class="p">,</span>
    <span class="n">use_scnets</span><span class="o">=</span><span class="n">use_scnets</span><span class="p">,</span>
    <span class="n">ntop_deg</span><span class="o">=</span><span class="n">ntop_deg</span><span class="p">,</span>
    <span class="n">ntop_deg_nodes</span><span class="o">=</span><span class="n">ntop_deg_nodes</span><span class="p">,</span>
    <span class="n">node_source</span><span class="o">=</span><span class="n">node_source</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">outputs</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">main_for_unaligned</span><span class="p">(</span>
    <span class="o">**</span><span class="n">came_inputs</span><span class="p">,</span>
    <span class="n">df_varmap</span><span class="o">=</span><span class="n">df_varmap</span><span class="p">,</span>
    <span class="n">df_varmap_1v1</span><span class="o">=</span><span class="n">df_varmap_1v1</span><span class="p">,</span>
    <span class="n">dataset_names</span><span class="o">=</span><span class="n">dsnames</span><span class="p">,</span>
    <span class="n">key_class1</span><span class="o">=</span><span class="n">key_class1</span><span class="p">,</span>
    <span class="n">key_class2</span><span class="o">=</span><span class="n">key_class2</span><span class="p">,</span>
    <span class="n">do_normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">keep_non1v1_feats</span><span class="o">=</span><span class="n">keep_non1v1_feats</span><span class="p">,</span>
    <span class="n">n_epochs</span><span class="o">=</span><span class="n">n_epochs</span><span class="p">,</span>
    <span class="n">resdir</span><span class="o">=</span><span class="n">resdir</span><span class="p">,</span>
    <span class="n">n_pass</span><span class="o">=</span><span class="n">n_pass</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
    <span class="n">plot_results</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dpair</span> <span class="o">=</span> <span class="n">outputs</span><span class="p">[</span><span class="s1">&#39;dpair&#39;</span><span class="p">]</span>
<span class="n">trainer</span> <span class="o">=</span> <span class="n">outputs</span><span class="p">[</span><span class="s1">&#39;trainer&#39;</span><span class="p">]</span>
<span class="n">h_dict</span> <span class="o">=</span> <span class="n">outputs</span><span class="p">[</span><span class="s1">&#39;h_dict&#39;</span><span class="p">]</span>
<span class="n">out_cell</span> <span class="o">=</span> <span class="n">outputs</span><span class="p">[</span><span class="s1">&#39;out_cell&#39;</span><span class="p">]</span>
<span class="n">predictor</span> <span class="o">=</span> <span class="n">outputs</span><span class="p">[</span><span class="s1">&#39;predictor&#39;</span><span class="p">]</span>


<span class="n">obs_ids1</span><span class="p">,</span> <span class="n">obs_ids2</span> <span class="o">=</span> <span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids1</span><span class="p">,</span> <span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids2</span>
<span class="n">obs</span> <span class="o">=</span> <span class="n">dpair</span><span class="o">.</span><span class="n">obs</span>
<span class="n">classes</span> <span class="o">=</span> <span class="n">predictor</span><span class="o">.</span><span class="n">classes</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[leiden] Time used:  0.2723 s
650 genes before taking unique
taking total of 501 unique differential expressed genes
400 genes before taking unique
taking total of 345 unique differential expressed genes
already exists:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs
already exists:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)
[*] Setting dataset names:
        0--&gt;Baron_human
        1--&gt;Baron_mouse
[*] Setting aligned features for observation nodes (self._features)
[*] Setting un-aligned features (`self._ov_adjs`) for making links connecting observation and variable nodes
[*] Setting adjacent matrix connecting variables from these 2 datasets (`self._vv_adj`)
Index([&#39;cell_ontology_class&#39;, &#39;cell_ontology_id&#39;, &#39;cell_type1&#39;, &#39;dataset_name&#39;,
       &#39;donor&#39;, &#39;latent_1&#39;, &#39;latent_10&#39;, &#39;latent_2&#39;, &#39;latent_3&#39;, &#39;latent_4&#39;,
       &#39;latent_5&#39;, &#39;latent_6&#39;, &#39;latent_7&#39;, &#39;latent_8&#39;, &#39;latent_9&#39;, &#39;library&#39;,
       &#39;organ&#39;, &#39;organism&#39;, &#39;platform&#39;, &#39;tSNE1&#39;, &#39;tSNE2&#39;],
      dtype=&#39;object&#39;)
Index([&#39;cell_ontology_class&#39;, &#39;cell_ontology_id&#39;, &#39;cell_type1&#39;, &#39;dataset_name&#39;,
       &#39;donor&#39;, &#39;latent_1&#39;, &#39;latent_10&#39;, &#39;latent_2&#39;, &#39;latent_3&#39;, &#39;latent_4&#39;,
       &#39;latent_5&#39;, &#39;latent_6&#39;, &#39;latent_7&#39;, &#39;latent_8&#39;, &#39;latent_9&#39;, &#39;library&#39;,
       &#39;organ&#39;, &#39;organism&#39;, &#39;platform&#39;, &#39;tSNE1&#39;, &#39;tSNE2&#39;, &#39;clust_lbs&#39;],
      dtype=&#39;object&#39;)
-------------------- Summary of the DGL-Heterograph --------------------
Graph(num_nodes={&#39;cell&#39;: 10455, &#39;gene&#39;: 6313},
      num_edges={(&#39;cell&#39;, &#39;express&#39;, &#39;gene&#39;): 3793858, (&#39;cell&#39;, &#39;self_loop_cell&#39;, &#39;cell&#39;): 10455, (&#39;cell&#39;, &#39;similar_to&#39;, &#39;cell&#39;): 66764, (&#39;gene&#39;, &#39;expressed_by&#39;, &#39;cell&#39;): 3793858, (&#39;gene&#39;, &#39;homolog_with&#39;, &#39;gene&#39;): 11915},
      metagraph=[(&#39;cell&#39;, &#39;gene&#39;, &#39;express&#39;), (&#39;cell&#39;, &#39;cell&#39;, &#39;self_loop_cell&#39;), (&#39;cell&#39;, &#39;cell&#39;, &#39;similar_to&#39;), (&#39;gene&#39;, &#39;cell&#39;, &#39;expressed_by&#39;), (&#39;gene&#39;, &#39;gene&#39;, &#39;homolog_with&#39;)])
second-order connection: False
self-loops for observation-nodes: True
self-loops for variable-nodes: True
DataPair with 10455 obs- and 6313 var-nodes
obs1 x var1 (Baron_human): 8569 x 3332
obs2 x var2 (Baron_mouse): 1886 x 2981
Dimensions of the obs-node-features: 547
a new directory made:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\_models
main directory: _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)
model directory: _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\_models
============== start training (device=&#39;cuda&#39;) ==============
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\Administrator\AppData\Roaming\Python\Python38\site-packages\torch\nn\modules\container.py:552: UserWarning: Setting attributes on ParameterDict is not supported.
  warnings.warn(&#34;Setting attributes on ParameterDict is not supported.&#34;)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 0000 | Train Acc: 0.0533 | Test: 0.0323 (max=0.0323) | AMI=-0.0000 | Time: 0.4114
Epoch 0005 | Train Acc: 0.2948 | Test: 0.4740 (max=0.4804) | AMI=0.0435 | Time: 0.1553
Epoch 0010 | Train Acc: 0.5011 | Test: 0.4899 (max=0.4899) | AMI=0.4180 | Time: 0.1315
Epoch 0015 | Train Acc: 0.4372 | Test: 0.6193 (max=0.6193) | AMI=0.4547 | Time: 0.1239
Epoch 0020 | Train Acc: 0.7186 | Test: 0.7174 (max=0.7174) | AMI=0.5456 | Time: 0.1194
Epoch 0025 | Train Acc: 0.7508 | Test: 0.7216 (max=0.7301) | AMI=0.5801 | Time: 0.1166
Epoch 0030 | Train Acc: 0.8419 | Test: 0.7937 (max=0.7937) | AMI=0.6283 | Time: 0.1149
Epoch 0035 | Train Acc: 0.8615 | Test: 0.8165 (max=0.8165) | AMI=0.6809 | Time: 0.1134
Epoch 0040 | Train Acc: 0.8743 | Test: 0.7773 (max=0.8165) | AMI=0.6177 | Time: 0.1124
Epoch 0045 | Train Acc: 0.8766 | Test: 0.7927 (max=0.8340) | AMI=0.6383 | Time: 0.1117
Epoch 0050 | Train Acc: 0.9152 | Test: 0.8208 (max=0.8340) | AMI=0.6644 | Time: 0.1110
Epoch 0055 | Train Acc: 0.9121 | Test: 0.8160 (max=0.8574) | AMI=0.6521 | Time: 0.1105
Epoch 0060 | Train Acc: 0.9420 | Test: 0.8627 (max=0.9067) | AMI=0.6900 | Time: 0.1101
Epoch 0065 | Train Acc: 0.9455 | Test: 0.8425 (max=0.9120) | AMI=0.6644 | Time: 0.1097
Epoch 0070 | Train Acc: 0.9531 | Test: 0.9067 (max=0.9120) | AMI=0.7347 | Time: 0.1096
Epoch 0075 | Train Acc: 0.9585 | Test: 0.9152 (max=0.9337) | AMI=0.7561 | Time: 0.1092
Epoch 0080 | Train Acc: 0.9639 | Test: 0.9268 (max=0.9337) | AMI=0.7646 | Time: 0.1091
Epoch 0085 | Train Acc: 0.9679 | Test: 0.9194 (max=0.9337) | AMI=0.7552 | Time: 0.1088
Epoch 0090 | Train Acc: 0.9722 | Test: 0.9290 (max=0.9358) | AMI=0.7685 | Time: 0.1088
Epoch 0095 | Train Acc: 0.9537 | Test: 0.9051 (max=0.9358) | AMI=0.7701 | Time: 0.1086
[current best] model weights backup
Epoch 0099 | Train Acc: 0.9673 | Test: 0.9099 (max=0.9358) | AMI=0.7497 | Time: 0.1088
[current best] model weights backup
Epoch 0100 | Train Acc: 0.9781 | Test: 0.9168 (max=0.9358) | AMI=0.7655 | Time: 0.1088
[current best] model weights backup
Epoch 0101 | Train Acc: 0.9718 | Test: 0.9300 (max=0.9358) | AMI=0.7725 | Time: 0.1091
[current best] model weights backup
Epoch 0104 | Train Acc: 0.9805 | Test: 0.9300 (max=0.9358) | AMI=0.7765 | Time: 0.1090
Epoch 0105 | Train Acc: 0.9798 | Test: 0.9231 (max=0.9358) | AMI=0.7519 | Time: 0.1090
[current best] model weights backup
Epoch 0106 | Train Acc: 0.9791 | Test: 0.9316 (max=0.9358) | AMI=0.7855 | Time: 0.1090
[current best] model weights backup
Epoch 0107 | Train Acc: 0.9807 | Test: 0.9300 (max=0.9358) | AMI=0.7866 | Time: 0.1090
[current best] model weights backup
Epoch 0109 | Train Acc: 0.9791 | Test: 0.9300 (max=0.9358) | AMI=0.7885 | Time: 0.1091
Epoch 0110 | Train Acc: 0.9793 | Test: 0.9210 (max=0.9358) | AMI=0.7842 | Time: 0.1090
Epoch 0115 | Train Acc: 0.9831 | Test: 0.9337 (max=0.9358) | AMI=0.7743 | Time: 0.1089
Epoch 0120 | Train Acc: 0.9835 | Test: 0.9295 (max=0.9358) | AMI=0.7763 | Time: 0.1088
[current best] model weights backup
Epoch 0123 | Train Acc: 0.9830 | Test: 0.9332 (max=0.9358) | AMI=0.8002 | Time: 0.1088
Epoch 0125 | Train Acc: 0.9828 | Test: 0.9014 (max=0.9358) | AMI=0.7312 | Time: 0.1088
model weights backup
Epoch 0129 | Train Acc: 0.9842 | Test: 0.9284 (max=0.9358) | AMI=0.7831 | Time: 0.1088
Epoch 0130 | Train Acc: 0.9821 | Test: 0.9364 (max=0.9364) | AMI=0.7930 | Time: 0.1088
Epoch 0135 | Train Acc: 0.9867 | Test: 0.9295 (max=0.9364) | AMI=0.7676 | Time: 0.1088
Epoch 0140 | Train Acc: 0.9837 | Test: 0.9252 (max=0.9401) | AMI=0.7799 | Time: 0.1087
Epoch 0145 | Train Acc: 0.9869 | Test: 0.9321 (max=0.9401) | AMI=0.7796 | Time: 0.1087
Epoch 0150 | Train Acc: 0.9882 | Test: 0.9247 (max=0.9401) | AMI=0.7689 | Time: 0.1086
[current best] model weights backup
Epoch 0152 | Train Acc: 0.9888 | Test: 0.9401 (max=0.9401) | AMI=0.8028 | Time: 0.1086
Epoch 0155 | Train Acc: 0.9893 | Test: 0.9374 (max=0.9401) | AMI=0.7850 | Time: 0.1086
Epoch 0160 | Train Acc: 0.9895 | Test: 0.9311 (max=0.9401) | AMI=0.7703 | Time: 0.1085
Epoch 0165 | Train Acc: 0.9894 | Test: 0.9337 (max=0.9401) | AMI=0.7990 | Time: 0.1084
[current best] model weights backup
Epoch 0167 | Train Acc: 0.9891 | Test: 0.9406 (max=0.9406) | AMI=0.8053 | Time: 0.1085
Epoch 0170 | Train Acc: 0.9895 | Test: 0.9369 (max=0.9406) | AMI=0.7956 | Time: 0.1085
model weights backup
Epoch 0172 | Train Acc: 0.9894 | Test: 0.9464 (max=0.9464) | AMI=0.7894 | Time: 0.1086
Epoch 0175 | Train Acc: 0.9897 | Test: 0.9427 (max=0.9464) | AMI=0.7962 | Time: 0.1085
[current best] model weights backup
Epoch 0178 | Train Acc: 0.9869 | Test: 0.9385 (max=0.9464) | AMI=0.8072 | Time: 0.1085
Epoch 0180 | Train Acc: 0.9893 | Test: 0.9417 (max=0.9464) | AMI=0.8058 | Time: 0.1085
Epoch 0185 | Train Acc: 0.9911 | Test: 0.9300 (max=0.9464) | AMI=0.7838 | Time: 0.1084
Epoch 0190 | Train Acc: 0.9908 | Test: 0.9364 (max=0.9464) | AMI=0.7876 | Time: 0.1084
Epoch 0195 | Train Acc: 0.9911 | Test: 0.9343 (max=0.9464) | AMI=0.7741 | Time: 0.1084
Epoch 0200 | Train Acc: 0.9895 | Test: 0.9369 (max=0.9464) | AMI=0.8005 | Time: 0.1083
Epoch 0205 | Train Acc: 0.9896 | Test: 0.9411 (max=0.9464) | AMI=0.7990 | Time: 0.1082
Epoch 0210 | Train Acc: 0.9907 | Test: 0.9183 (max=0.9464) | AMI=0.7845 | Time: 0.1082
model weights backup
Epoch 0215 | Train Acc: 0.9916 | Test: 0.9343 (max=0.9464) | AMI=0.7855 | Time: 0.1082
Epoch 0220 | Train Acc: 0.9918 | Test: 0.9390 (max=0.9475) | AMI=0.7800 | Time: 0.1082
Epoch 0225 | Train Acc: 0.9924 | Test: 0.9422 (max=0.9475) | AMI=0.7965 | Time: 0.1082
Epoch 0230 | Train Acc: 0.9932 | Test: 0.9443 (max=0.9475) | AMI=0.8003 | Time: 0.1082
[current best] model weights backup
Epoch 0235 | Train Acc: 0.9915 | Test: 0.9422 (max=0.9475) | AMI=0.8147 | Time: 0.1082
Epoch 0240 | Train Acc: 0.9900 | Test: 0.9284 (max=0.9475) | AMI=0.8111 | Time: 0.1081
Epoch 0245 | Train Acc: 0.9925 | Test: 0.9374 (max=0.9475) | AMI=0.7860 | Time: 0.1081
[current best] model weights backup
Epoch 0249 | Train Acc: 0.9914 | Test: 0.9475 (max=0.9475) | AMI=0.8221 | Time: 0.1081
Epoch 0250 | Train Acc: 0.9925 | Test: 0.9332 (max=0.9475) | AMI=0.7768 | Time: 0.1081
Epoch 0255 | Train Acc: 0.9936 | Test: 0.9385 (max=0.9475) | AMI=0.7926 | Time: 0.1080
model weights backup
Epoch 0258 | Train Acc: 0.9939 | Test: 0.9401 (max=0.9475) | AMI=0.7941 | Time: 0.1080
Epoch 0260 | Train Acc: 0.9946 | Test: 0.9449 (max=0.9475) | AMI=0.8087 | Time: 0.1080
Epoch 0265 | Train Acc: 0.9938 | Test: 0.9417 (max=0.9475) | AMI=0.8002 | Time: 0.1080
Epoch 0270 | Train Acc: 0.9935 | Test: 0.9390 (max=0.9491) | AMI=0.7870 | Time: 0.1079
[current best] model weights backup
Epoch 0275 | Train Acc: 0.9945 | Test: 0.9449 (max=0.9491) | AMI=0.8241 | Time: 0.1079
Epoch 0280 | Train Acc: 0.9951 | Test: 0.9454 (max=0.9491) | AMI=0.8111 | Time: 0.1079
Epoch 0285 | Train Acc: 0.9933 | Test: 0.9459 (max=0.9491) | AMI=0.8047 | Time: 0.1078
Epoch 0290 | Train Acc: 0.9954 | Test: 0.9486 (max=0.9491) | AMI=0.8155 | Time: 0.1078
Epoch 0295 | Train Acc: 0.9950 | Test: 0.9454 (max=0.9491) | AMI=0.8052 | Time: 0.1078
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\cluster_index.png
states loaded from: _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\_models\weights_epoch275.pt
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\Administrator\AppData\Roaming\Python\Python38\site-packages\torch\nn\modules\container.py:552: UserWarning: Setting attributes on ParameterDict is not supported.
  warnings.warn(&#34;Setting attributes on ParameterDict is not supported.&#34;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
object saved into:
         _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\datapair_init.pickle
Re-order the rows
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\contingency_matrix(acc94.5%).png
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\contingency_matrix-train.png
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\heatmap_probas.pdf
</pre></div></div>
</div>
<div class="section" id="Load-other-checkpoint-(optional)">
<h3>Load other checkpoint (optional)<a class="headerlink" href="#Load-other-checkpoint-(optional)" title="Permalink to this headline">¶</a></h3>
<p>You can load other model checkpoint if the default model is not satisfying.</p>
<p>For example, load the last checkpoint and compute the results of it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">outputs</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">gather_came_results</span><span class="p">(</span>
    <span class="n">dpair</span><span class="p">,</span>
    <span class="n">trainer</span><span class="p">,</span>
    <span class="n">classes</span><span class="o">=</span><span class="n">classes</span><span class="p">,</span>
    <span class="n">keys</span><span class="o">=</span><span class="p">(</span><span class="n">key_class1</span><span class="p">,</span> <span class="n">key_class2</span><span class="p">),</span>
    <span class="n">keys_compare</span><span class="o">=</span><span class="p">(</span><span class="n">key_class1</span><span class="p">,</span> <span class="n">key_class2</span><span class="p">),</span>
    <span class="n">resdir</span><span class="o">=</span><span class="n">resdir</span><span class="p">,</span>
    <span class="n">checkpoint</span><span class="o">=</span><span class="s1">&#39;last&#39;</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>You can get all saved checkpoint numbers by:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">came</span><span class="o">.</span><span class="n">get_checkpoint_list</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;_models&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[100,
 101,
 104,
 106,
 107,
 109,
 123,
 129,
 152,
 167,
 172,
 178,
 215,
 235,
 249,
 258,
 275,
 299,
 99]
</pre></div></div>
</div>
</div>
<div class="section" id="Plot-the-contingency-matrix-for-query-dataset">
<h3>Plot the contingency matrix for query dataset<a class="headerlink" href="#Plot-the-contingency-matrix-for-query-dataset" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># contingency matrix for query dataset</span>
<span class="n">y_true</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">][</span><span class="n">obs_ids2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">y_pred</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="s1">&#39;predicted&#39;</span><span class="p">][</span><span class="n">obs_ids2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

<span class="n">ax</span><span class="p">,</span> <span class="n">contmat</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">plot_contingency_mat</span><span class="p">(</span>
    <span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">,</span> <span class="n">norm_axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">order_rows</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">order_cols</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pl</span><span class="o">.</span><span class="n">_save_with_adjust</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="p">,</span> <span class="n">figdir</span> <span class="o">/</span> <span class="s1">&#39;contingency_mat.png&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">figure</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\contingency_mat.png
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_17_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_17_1.png" />
</div>
</div>
</div>
<div class="section" id="Plot-heatmap-of-predicted-probabilities">
<h3>Plot heatmap of predicted probabilities<a class="headerlink" href="#Plot-heatmap-of-predicted-probabilities" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">name_label</span> <span class="o">=</span> <span class="s1">&#39;celltype&#39;</span>
<span class="n">cols_anno</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="s1">&#39;predicted&#39;</span><span class="p">][:]</span>
<span class="n">df_probs</span> <span class="o">=</span> <span class="n">obs</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">classes</span><span class="p">)]</span>


<span class="n">gs</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">wrapper_heatmap_scores</span><span class="p">(</span>
        <span class="n">df_probs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">obs_ids2</span><span class="p">],</span> <span class="n">obs</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">obs_ids2</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">col_label</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="n">col_pred</span><span class="o">=</span><span class="s1">&#39;predicted&#39;</span><span class="p">,</span>
        <span class="n">n_subsample</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>  <span class="c1"># sampling 50 cells for each group</span>
        <span class="n">cmap_heat</span><span class="o">=</span><span class="s1">&#39;magma_r&#39;</span><span class="p">,</span>
        <span class="n">fp</span><span class="o">=</span><span class="n">figdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;heatmap_probas.pdf&#39;</span>
    <span class="p">)</span>

<span class="n">gs</span><span class="o">.</span><span class="n">figure</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\heatmap_probas.pdf
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_19_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_19_1.png" />
</div>
</div>
</div>
</div>
<div class="section" id="2-Further-analysis">
<h2>2 Further analysis<a class="headerlink" href="#2-Further-analysis" title="Permalink to this headline">¶</a></h2>
<p>By default, CAME will use the <strong>last</strong> layer of hidden states, as the embeddings, to produce cell- and gene-UMAP.</p>
<blockquote>
<div><p>You can also load ALL of the model hidden states that have been seved during CAME’s default pipeline:</p>
</div></blockquote>
<div class="highlight-py notranslate"><div class="highlight"><pre><span></span><span class="n">hidden_list</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">load_hidden_states</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;hidden_list.h5&#39;</span><span class="p">)</span>
<span class="n">hidden_list</span>  <span class="c1"># a list of dicts</span>
<span class="n">h_dict</span> <span class="o">=</span> <span class="n">hidden_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span> <span class="c1"># the last layer of hidden states</span>
</pre></div>
</div>
<p>Make AnnData objects, storing only the CAME-embeddings and annotations, for cells and genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adt</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">make_adata</span><span class="p">(</span><span class="n">h_dict</span><span class="p">[</span><span class="s1">&#39;cell&#39;</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">dpair</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">assparse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">gadt</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">make_adata</span><span class="p">(</span><span class="n">h_dict</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">],</span> <span class="n">obs</span><span class="o">=</span><span class="n">dpair</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">assparse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># adt.write(resdir / &#39;adt_hidden_cell.h5ad&#39;)</span>
<span class="c1"># gadt.write_h5ad(resdir / &#39;adt_hidden_gene.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
adding columns to `adata.obs` (ignore_index=True):
original_name, dataset, REF, celltype, predicted, max_probs, is_right, pancreatic acinar cell, type B pancreatic cell, pancreatic D cell, pancreatic stellate cell, pancreatic ductal cell, pancreatic A cell, pancreatic epsilon cell, pancreatic PP cell, endothelial cell, macrophage, Schwann cell, mast cell, T cell, done!
adding columns to `adata.obs` (ignore_index=True):
name, dataset, done!
</pre></div></div>
</div>
<div class="section" id="2.1-UMAP-of-cell-embeddings">
<h3>2.1 UMAP of cell embeddings<a class="headerlink" href="#2.1-UMAP-of-cell-embeddings" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi_save</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adt</span><span class="p">)</span>
<span class="c1"># sc.pl.umap(adt, color=[&#39;dataset&#39;, &#39;celltype&#39;], ncols=1)</span>

<span class="n">ftype</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;.svg&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;-dataset</span><span class="si">{</span><span class="n">ftype</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adt</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;celltype&#39;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;-ctype</span><span class="si">{</span><span class="n">ftype</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;dataset&#39; as categorical
... storing &#39;REF&#39; as categorical
... storing &#39;celltype&#39; as categorical
... storing &#39;predicted&#39; as categorical
WARNING: saving figure to file _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap-dataset.pdf
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_23_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_23_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap-ctype.pdf
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_23_3.png" src="../_images/tut_notebooks_getting_started_pipeline_un_23_3.png" />
</div>
</div>
<p>Store UMAP coordinates:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">obs_umap</span> <span class="o">=</span> <span class="n">adt</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span>
<span class="n">obs</span><span class="p">[</span><span class="s1">&#39;UMAP1&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs_umap</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">obs</span><span class="p">[</span><span class="s1">&#39;UMAP2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs_umap</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">obs</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;obs.csv&#39;</span><span class="p">)</span>
<span class="n">adt</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;adt_hidden_cell.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Setting UMAP to the original adata</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">adata1</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs_umap</span><span class="p">[</span><span class="n">obs_ids1</span><span class="p">]</span>
<span class="n">adata2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">obs_umap</span><span class="p">[</span><span class="n">obs_ids2</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.2-UMAP-of-genes">
<h3>2.2 UMAP of genes<a class="headerlink" href="#2.2-UMAP-of-genes" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">gadt</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X&#39;</span><span class="p">)</span>

<span class="c1"># gadt = pp.make_adata(h_dict[&#39;gene&#39;], obs=dpair.var.iloc[:, :2], assparse=False, ignore_index=True)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gadt</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gadt</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;name&#39; as categorical
... storing &#39;dataset&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_29_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_29_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># joint gene module extraction</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">gadt</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=.</span><span class="mi">8</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;module&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gadt</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;module&#39;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s1">&#39;tab20b&#39;</span><span class="p">)</span>

<span class="c1"># backup results is needed</span>
<span class="c1"># gadt.write(resdir / &#39;adt_hidden_gene.h5ad&#39;)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_30_0.png" src="../_images/tut_notebooks_getting_started_pipeline_un_30_0.png" />
</div>
</div>
<p>Visualize gene UMAPs:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># gadt.obs_names = gadt.obs_names.astype(str)</span>
<span class="n">gadt1</span><span class="p">,</span> <span class="n">gadt2</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">bisplit_adata</span><span class="p">(</span><span class="n">gadt</span><span class="p">,</span> <span class="s1">&#39;dataset&#39;</span><span class="p">,</span> <span class="n">dsnames</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">reset_index_by</span><span class="o">=</span><span class="s1">&#39;name&#39;</span><span class="p">)</span>

<span class="n">color_by</span> <span class="o">=</span> <span class="s1">&#39;module&#39;</span>
<span class="n">palette</span> <span class="o">=</span> <span class="s1">&#39;tab20b&#39;</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gadt1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color_by</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">edges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">edges_width</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
           <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
           <span class="n">save</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">color_by</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">dsnames</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gadt2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color_by</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">edges</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">edges_width</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
           <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span>
           <span class="n">save</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{</span><span class="n">color_by</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">dsnames</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap_module-Baron_human.pdf
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_32_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_32_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING: saving figure to file _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap_module-Baron_human.pdf
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_32_3.png" src="../_images/tut_notebooks_getting_started_pipeline_un_32_3.png" />
</div>
</div>
<p>Compute the link-weights (embedding similarity) between homologous gene pairs:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df_var_links</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">weight_linked_vars</span><span class="p">(</span>
    <span class="n">gadt</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">dpair</span><span class="o">.</span><span class="n">_vv_adj</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="n">dpair</span><span class="o">.</span><span class="n">get_vnode_names</span><span class="p">(),</span>
    <span class="n">matric</span><span class="o">=</span><span class="s1">&#39;cosine&#39;</span><span class="p">,</span> <span class="n">index_names</span><span class="o">=</span><span class="n">dsnames</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">gadt1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;adt_hidden_gene1.h5ad&#39;</span><span class="p">)</span>
<span class="n">gadt2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;adt_hidden_gene2.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.3-Gene-expression-profiles-(for-each-cell-type)-on-gene-UMAP">
<h3>2.3 Gene-expression-profiles (for each cell type) on gene UMAP<a class="headerlink" href="#2.3-Gene-expression-profiles-(for-each-cell-type)-on-gene-UMAP" title="Permalink to this headline">¶</a></h3>
<p>Compute average expressions for each cell type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># averaged expressions</span>
<span class="n">avg_expr1</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">group_mean_adata</span><span class="p">(</span>
    <span class="n">adatas</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">groupby</span><span class="o">=</span><span class="n">key_class1</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">dpair</span><span class="o">.</span><span class="n">vnode_names1</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">avg_expr2</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">group_mean_adata</span><span class="p">(</span>
    <span class="n">adatas</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">groupby</span><span class="o">=</span><span class="n">key_class2</span> <span class="k">if</span> <span class="n">key_class2</span> <span class="k">else</span> <span class="s1">&#39;predicted&#39;</span><span class="p">,</span>
    <span class="n">features</span><span class="o">=</span><span class="n">dpair</span><span class="o">.</span><span class="n">vnode_names2</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing averages grouped by cell_ontology_class
Computing averages grouped by cell_ontology_class
</pre></div></div>
</div>
<p>plot cell type gene-profiles (plot all the cell types) on UMAP</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plkwds</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdYlBu_r&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">2.5</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">do_zscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
              <span class="n">axscale</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">with_cbar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">fig1</span><span class="p">,</span> <span class="n">axs1</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">adata_embed_with_values</span><span class="p">(</span>
    <span class="n">gadt1</span><span class="p">,</span> <span class="n">avg_expr1</span><span class="p">,</span>
    <span class="n">fp</span><span class="o">=</span><span class="n">figdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;umap_exprAvgs-</span><span class="si">{</span><span class="n">dsnames</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">-all.png&#39;</span><span class="p">,</span>
    <span class="o">**</span><span class="n">plkwds</span><span class="p">)</span>
<span class="n">fig2</span><span class="p">,</span> <span class="n">axs2</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">adata_embed_with_values</span><span class="p">(</span>
    <span class="n">gadt2</span><span class="p">,</span> <span class="n">avg_expr2</span><span class="p">,</span>
    <span class="n">fp</span><span class="o">=</span><span class="n">figdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;umap_exprAvgs-</span><span class="si">{</span><span class="n">dsnames</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">-all.png&#39;</span><span class="p">,</span>
    <span class="o">**</span><span class="n">plkwds</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap_exprAvgs-Baron_human-all.png
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\umap_exprAvgs-Baron_mouse-all.png
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_39_0.png" src="../_images/tut_notebooks_getting_started_pipeline_un_39_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">fig2</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_40_0.png" src="../_images/tut_notebooks_getting_started_pipeline_un_40_0.png" />
</div>
</div>
</div>
<div class="section" id="2.4-Abstracted-graph">
<h3>2.4 Abstracted graph<a class="headerlink" href="#2.4-Abstracted-graph" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">norm_ov</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;zs&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="n">cut_ov</span> <span class="o">=</span> <span class="mf">0.</span>

<span class="n">groupby_var</span> <span class="o">=</span> <span class="s1">&#39;module&#39;</span>
<span class="n">obs_labels1</span><span class="p">,</span> <span class="n">obs_labels2</span> <span class="o">=</span> <span class="n">adt</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">][</span><span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids1</span><span class="p">],</span> \
                           <span class="n">adt</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">][</span><span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids2</span><span class="p">]</span>
<span class="n">var_labels1</span><span class="p">,</span> <span class="n">var_labels2</span> <span class="o">=</span> <span class="n">gadt1</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby_var</span><span class="p">],</span> <span class="n">gadt2</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">groupby_var</span><span class="p">]</span>

<span class="n">sp1</span><span class="p">,</span> <span class="n">sp2</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;mouse&#39;</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">make_abstracted_graph</span><span class="p">(</span>
    <span class="n">obs_labels1</span><span class="p">,</span> <span class="n">obs_labels2</span><span class="p">,</span>
    <span class="n">var_labels1</span><span class="p">,</span> <span class="n">var_labels2</span><span class="p">,</span>
    <span class="n">avg_expr1</span><span class="p">,</span> <span class="n">avg_expr2</span><span class="p">,</span>
    <span class="n">df_var_links</span><span class="p">,</span>
    <span class="n">tags_obs</span><span class="o">=</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sp1</span><span class="si">}</span><span class="s1"> &#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sp2</span><span class="si">}</span><span class="s1"> &#39;</span><span class="p">),</span>
    <span class="n">tags_var</span><span class="o">=</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sp1</span><span class="si">}</span><span class="s1"> module &#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sp2</span><span class="si">}</span><span class="s1"> module &#39;</span><span class="p">),</span>
    <span class="n">cut_ov</span><span class="o">=</span><span class="n">cut_ov</span><span class="p">,</span>
    <span class="n">norm_mtd_ov</span><span class="o">=</span><span class="n">norm_ov</span><span class="p">,</span>
<span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3332 2981
Edges with weights lower than 0 were cut out.
Edges with weights lower than 0 were cut out.
2801
2801 2801
---&gt; aggregating edges...
unique labels of rows: [&#39;5&#39; &#39;1&#39; &#39;2&#39; &#39;4&#39; &#39;6&#39; &#39;0&#39; &#39;9&#39; &#39;3&#39; &#39;7&#39; &#39;8&#39; &#39;10&#39;]
unique labels of columns: [&#39;6&#39; &#39;2&#39; &#39;0&#39; &#39;7&#39; &#39;8&#39; &#39;9&#39; &#39;3&#39; &#39;5&#39; &#39;1&#39; &#39;4&#39; &#39;10&#39;]
grouping elements (edges)
shape of the one-hot-labels: (3332, 11) (2981, 11)
Re-order the rows
Re-order the columns
Re-order the columns
Re-order the columns
2801
2801 2801
---&gt; aggregating edges...
unique labels of rows: [&#39;5&#39; &#39;1&#39; &#39;2&#39; &#39;4&#39; &#39;6&#39; &#39;0&#39; &#39;9&#39; &#39;3&#39; &#39;7&#39; &#39;8&#39; &#39;10&#39;]
unique labels of columns: [&#39;6&#39; &#39;2&#39; &#39;0&#39; &#39;7&#39; &#39;8&#39; &#39;9&#39; &#39;3&#39; &#39;5&#39; &#39;1&#39; &#39;4&#39; &#39;10&#39;]
grouping elements (edges)
shape of the one-hot-labels: (3332, 11) (2981, 11)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="sd">&#39;&#39;&#39; visualization &#39;&#39;&#39;</span>
<span class="n">fp_abs</span> <span class="o">=</span> <span class="n">figdir</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;abstracted_graph-</span><span class="si">{</span><span class="n">groupby_var</span><span class="si">}</span><span class="s1">-cut</span><span class="si">{</span><span class="n">cut_ov</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">norm_ov</span><span class="si">}</span><span class="s1">.pdf&#39;</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">plot_multipartite_graph</span><span class="p">(</span>
    <span class="n">g</span><span class="p">,</span> <span class="n">edge_scale</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">),</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">fp</span><span class="o">=</span><span class="n">fp_abs</span><span class="p">)</span>  <span class="c1"># nodelist=nodelist,</span>

<span class="n">ax</span><span class="o">.</span><span class="n">figure</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[13, 11, 11, 12]
figure has been saved into:
        _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\figs\abstracted_graph-module-cut0.0-zs.pdf
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/tut_notebooks_getting_started_pipeline_un_43_1.png" src="../_images/tut_notebooks_getting_started_pipeline_un_43_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># save abstracted-graph object</span>
<span class="n">came</span><span class="o">.</span><span class="n">save_pickle</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">resdir</span> <span class="o">/</span> <span class="s1">&#39;abs_graph.pickle&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
object saved into:
         _temp\(&#39;Baron_human&#39;, &#39;Baron_mouse&#39;)-(12-16 18.36.34)\abs_graph.pickle
</pre></div></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="getting_started_pipeline_aligned.html" class="btn btn-neutral float-right" title="Getting started (for datasets with aligned features)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2022, Academy of Mathematics and Systems Science, CAS..

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>